shell/platform/android/io/flutter/embedding/engine/FlutterEngineAndroidLifecycle.java

```diff
@@ -0,0 +1,126 @@
+package io.flutter.embedding.engine;
+
+import android.arch.lifecycle.DefaultLifecycleObserver;
+import android.arch.lifecycle.Lifecycle;
+import android.arch.lifecycle.LifecycleObserver;
+import android.arch.lifecycle.LifecycleOwner;
+import android.arch.lifecycle.LifecycleRegistry;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+/**
+ * Android {@link Lifecycle} that is owned by a {@link FlutterEngine}.
+ * <p>
+ * {@code FlutterEngineAndroidLifecycle} exists so that {@code FlutterPlugin}s can monitor Android
+ * lifecycle events. When the associated {@link FlutterEngine} is running in an {@code Activity},
+ * that {@code Activity}'s {@link Lifecycle} can be set as the {@code backingLifecycle} of this
+ * class, allowing all Flutter plugins to receive the {@code Activity}'s lifecycle events. Likewise,
+ * when the associated {@link FlutterEngine} is running in a {@code Service}, that {@code Service}'s
+ * {@link Lifecycle} can be set as the {@code backingLifecycle}.
+ * <p>
+ * Sometimes a {@link FlutterEngine} exists in a non-lifecycle location, e.g., an {@code Application},
+ * {@code ContentProvider}, or {@code BroadcastReceiver}. In these cases, this lifecycle reports
+ * itself in the {@link Lifecycle.State#CREATED} state.
+ * <p>
+ * Regardless of what happens to a backing {@code Activity} or @{code Service}, this lifecycle
```

Comment:
It's unexpected to me that `onDestroyed` would be special cased to the `FlutterEngine` specifically and that it's impossible to get an `onDestroyed` callback for the actual underlying lifecycle. Any way we could propagate this information life normal without breaking any kind of Engine contract? Add in an `onEngineDestroyed` method just for this class to handle the final engine destruction?

Reply:
The contract that we're trying not to violate is the `Lifecycle` contract, not the `FlutterEngine` contract: https://developer.android.com/reference/android/arch/lifecycle/Lifecycle.State#destroyed

The unique thing about `FlutterEngine` is that its backing `Lifecycle` can change over time. It goes from `Activity` to `Service` to nothing to `Activity` again. There is nothing preventing us from forwarding an `Activity`'s `ON_DESTROY` event, but then any other code that is designed to understand `Lifecycle` events will have undefined behavior when the `FlutterEngine`'s `Lifecycle` turns right around and emits an `ON_CREATED` event. Additionally, it is also conceptually incorrect to send that `ON_DESTROY` event because even though the `Activity` has been destroyed, the `FlutterEngine` remains, and at the end of the day this `Lifecycle` is truly owned by the `FlutterEngine`, it just temporarily takes on the `Lifecycle` of various Android components.

Does that make sense?